cms/bundles/signal_handlers.py

```diff
@@ -39,3 +50,128 @@ def handle_bundle_publication(instance: Bundle, **kwargs: Any) -> None:
+    """Handle when a dataset is added to a bundle."""
+    # Only proceed if the dataset is being created (i.e. is not getting updated) and the bundle has a bundle_api_id
+    if not created or not instance.parent.bundle_api_id:
+        return
+
+    client = BundleAPIClient()
+
+    try:
+        # Add the content item to the bundle
+        content_item = build_content_item_for_dataset(instance.dataset)
+
+        response = client.add_content_to_bundle(instance.parent.bundle_api_id, content_item)
+
+        # Extract content_id from the response
+        content_id = extract_content_id_from_bundle_response(response, instance.dataset)
+
+        if content_id:
+            instance.content_api_id = content_id
+            instance.save(update_fields=["content_api_id"])
+            logger.info(
+                "Added content %s to bundle %s in Dataset API with ID %s",
+                instance.dataset.namespace,
+                instance.parent.pk,
+                content_id,
+            )
+        else:
+            logger.error("Could not find content_id in response for bundle %s", instance.parent.pk)
+
+    except BundleAPIClientError as e:
+        logger.error("Failed to add content to bundle %s in Dataset API: %s", instance.parent.pk, e)
+
+
+@receiver(post_delete, sender=BundleDataset)
+@ons_bundle_api_enabled
+def handle_bundle_dataset_removed(instance: BundleDataset, **kwargs: Any) -> None:
+    """Handle when a dataset is removed from a bundle."""
+    if not instance.parent.bundle_api_id or not instance.content_api_id:
+        return
+
+    client = BundleAPIClient()
+
+    try:
+        content_id = instance.content_api_id
+        client.delete_content_from_bundle(instance.parent.bundle_api_id, content_id)
+        logger.info("Deleted content %s from bundle %s in Dataset API", content_id, instance.parent.pk)
+    except BundleAPIClientError as e:
+        logger.error("Failed to delete content %s from bundle %s in Dataset API: %s", content_id, instance.parent.pk, e)
+        # Don't raise the exception to avoid breaking the deletion process
+
+
+@receiver(post_delete, sender=Bundle)
+@ons_bundle_api_enabled
+def handle_bundle_deletion(instance: Bundle, **kwargs: Any) -> None:
+    """Handle when a bundle is deleted."""
+    if instance.bundle_api_id:
+        client = BundleAPIClient()
+
+        try:
+            client.delete_bundle(instance.bundle_api_id)
+            logger.info("Deleted bundle %s from Dataset API", instance.pk)
+
+        except BundleAPIClientError as e:
+            logger.error("Failed to delete bundle %s from Dataset API: %s", instance.pk, e)
+
+
+def _sync_preview_teams_with_bundle_api(bundle: Bundle) -> None:
+    """Helper function to sync a bundle's preview teams with the Bundle API."""
+    if not bundle.bundle_api_id:
+        # This happens when no datasets have been added to the bundle yet
+        logger.info("Bundle %s does not have an API ID. Skipping preview team sync.", bundle.pk)
```

Comment:
question: do we need to log this? Having a no-dataset bundle is an acceptable use case and I wonder whether this adds noise where it should not? Plus the other handlers do not do this when there is no bundle_api_id 🤔 

cc @MebinAbraham

Reply:
I went with "log as much as possible until we have full integration with the live API", albeit you're right that I didn't do this in other handlers, so can either remove it from here or standardise it across handlers.